<!DOCTYPE html>
<html>
<head>
  <meta name="generator" content=
  "HTML Tidy for HTML5 for Linux version 5.0.0.RC1">
  <meta content="text/html; charset=utf-8" http-equiv=
  "content-type">
  <style type="text/css">
  ol.lst-kix_i49qhjyjm8rs-6.start{counter-reset:lst-ctn-kix_i49qhjyjm8rs-6 0}.lst-kix_i49qhjyjm8rs-1>li:before{content:"" counter(lst-ctn-kix_i49qhjyjm8rs-1,lower-latin) ". "}.lst-kix_i49qhjyjm8rs-3>li:before{content:"" counter(lst-ctn-kix_i49qhjyjm8rs-3,decimal) ". "}.lst-kix_i49qhjyjm8rs-0>li{counter-increment:lst-ctn-kix_i49qhjyjm8rs-0}.lst-kix_i49qhjyjm8rs-2>li:before{content:"" counter(lst-ctn-kix_i49qhjyjm8rs-2,lower-roman) ". "}.lst-kix_i49qhjyjm8rs-6>li:before{content:"" counter(lst-ctn-kix_i49qhjyjm8rs-6,decimal) ". "}.lst-kix_i49qhjyjm8rs-5>li:before{content:"" counter(lst-ctn-kix_i49qhjyjm8rs-5,lower-roman) ". "}.lst-kix_i49qhjyjm8rs-4>li:before{content:"" counter(lst-ctn-kix_i49qhjyjm8rs-4,lower-latin) ". "}.lst-kix_i49qhjyjm8rs-1>li{counter-increment:lst-ctn-kix_i49qhjyjm8rs-1}.lst-kix_i49qhjyjm8rs-7>li{counter-increment:lst-ctn-kix_i49qhjyjm8rs-7}.lst-kix_i49qhjyjm8rs-0>li:before{content:"" counter(lst-ctn-kix_i49qhjyjm8rs-0,decimal) ". "}.lst-kix_pkhhiafczwwt-6>li:before{content:"\0025cf  "}ol.lst-kix_i49qhjyjm8rs-0.start{counter-reset:lst-ctn-kix_i49qhjyjm8rs-0 0}.lst-kix_pkhhiafczwwt-7>li:before{content:"\0025cb  "}.lst-kix_pkhhiafczwwt-8>li:before{content:"\0025a0  "}ol.lst-kix_i49qhjyjm8rs-7.start{counter-reset:lst-ctn-kix_i49qhjyjm8rs-7 0}ol.lst-kix_i49qhjyjm8rs-3.start{counter-reset:lst-ctn-kix_i49qhjyjm8rs-3 0}.lst-kix_i49qhjyjm8rs-6>li{counter-increment:lst-ctn-kix_i49qhjyjm8rs-6}ol.lst-kix_i49qhjyjm8rs-1.start{counter-reset:lst-ctn-kix_i49qhjyjm8rs-1 0}.lst-kix_i49qhjyjm8rs-3>li{counter-increment:lst-ctn-kix_i49qhjyjm8rs-3}ol.lst-kix_i49qhjyjm8rs-6{list-style-type:none}ol.lst-kix_i49qhjyjm8rs-5{list-style-type:none}ol.lst-kix_i49qhjyjm8rs-8{list-style-type:none}ol.lst-kix_i49qhjyjm8rs-7{list-style-type:none}ol.lst-kix_i49qhjyjm8rs-4.start{counter-reset:lst-ctn-kix_i49qhjyjm8rs-4 0}.lst-kix_i49qhjyjm8rs-4>li{counter-increment:lst-ctn-kix_i49qhjyjm8rs-4}.lst-kix_i49qhjyjm8rs-8>li{counter-increment:lst-ctn-kix_i49qhjyjm8rs-8}.lst-kix_pkhhiafczwwt-5>li:before{content:"\0025a0  "}.lst-kix_pkhhiafczwwt-4>li:before{content:"\0025cb  "}ul.lst-kix_pkhhiafczwwt-1{list-style-type:none}.lst-kix_pkhhiafczwwt-2>li:before{content:"\0025a0  "}.lst-kix_i49qhjyjm8rs-5>li{counter-increment:lst-ctn-kix_i49qhjyjm8rs-5}ul.lst-kix_pkhhiafczwwt-2{list-style-type:none}.lst-kix_i49qhjyjm8rs-2>li{counter-increment:lst-ctn-kix_i49qhjyjm8rs-2}.lst-kix_pkhhiafczwwt-1>li:before{content:"\0025cb  "}.lst-kix_pkhhiafczwwt-3>li:before{content:"\0025cf  "}ol.lst-kix_i49qhjyjm8rs-2.start{counter-reset:lst-ctn-kix_i49qhjyjm8rs-2 0}ul.lst-kix_pkhhiafczwwt-0{list-style-type:none}ul.lst-kix_pkhhiafczwwt-5{list-style-type:none}ul.lst-kix_pkhhiafczwwt-6{list-style-type:none}ul.lst-kix_pkhhiafczwwt-3{list-style-type:none}ul.lst-kix_pkhhiafczwwt-4{list-style-type:none}ol.lst-kix_i49qhjyjm8rs-2{list-style-type:none}ol.lst-kix_i49qhjyjm8rs-1{list-style-type:none}ul.lst-kix_pkhhiafczwwt-7{list-style-type:none}ol.lst-kix_i49qhjyjm8rs-4{list-style-type:none}ul.lst-kix_pkhhiafczwwt-8{list-style-type:none}ol.lst-kix_i49qhjyjm8rs-3{list-style-type:none}.lst-kix_pkhhiafczwwt-0>li:before{content:"\0025cf  "}ol.lst-kix_i49qhjyjm8rs-0{list-style-type:none}ol.lst-kix_i49qhjyjm8rs-8.start{counter-reset:lst-ctn-kix_i49qhjyjm8rs-8 0}ol.lst-kix_i49qhjyjm8rs-5.start{counter-reset:lst-ctn-kix_i49qhjyjm8rs-5 0}.lst-kix_i49qhjyjm8rs-7>li:before{content:"" counter(lst-ctn-kix_i49qhjyjm8rs-7,lower-latin) ". "}.lst-kix_i49qhjyjm8rs-8>li:before{content:"" counter(lst-ctn-kix_i49qhjyjm8rs-8,lower-roman) ". "}ol{margin:0;padding:0}table td,table th{padding:0}.c8{color:#000000;font-weight:700;text-decoration:none;vertical-align:baseline;font-size:11pt;font-family:"Arial";font-style:normal}.c0{color:#000000;font-weight:400;text-decoration:none;vertical-align:baseline;font-size:11pt;font-family:"Arial";font-style:normal}.c9{color:#000000;font-weight:700;text-decoration:none;vertical-align:baseline;font-size:14pt;font-family:"Arial";font-style:normal}.c1{padding-top:0pt;padding-bottom:0pt;line-height:1.15;orphans:2;widows:2;text-align:left}.c17{color:#000000;text-decoration:none;vertical-align:baseline;font-size:16pt;font-family:"Arial";font-style:normal}.c14{color:#000000;text-decoration:none;vertical-align:baseline;font-size:12pt;font-family:"Arial";font-style:normal}.c6{padding-top:0pt;padding-bottom:0pt;line-height:1.15;page-break-after:avoid;text-align:left}.c22{padding-top:0pt;padding-bottom:0pt;line-height:1.15;text-align:right}.c7{padding-top:0pt;padding-bottom:0pt;line-height:1.0;text-align:left}.c11{background-color:#ffffff;max-width:468pt;padding:72pt 72pt 72pt 72pt}.c12{border:1px solid black;margin:5px}.c21{margin-left:72pt;padding-left:0pt}.c18{margin-left:36pt;padding-left:0pt}.c3{color:#1155cc;text-decoration:underline}.c19{padding:0;margin:0}.c2{color:inherit;text-decoration:inherit}.c15{orphans:2;widows:2}.c20{vertical-align:super}.c16{font-weight:400}.c10{page-break-after:avoid}.c4{height:11pt}.c5{font-style:italic}.c13{font-weight:700}.title{padding-top:0pt;color:#000000;font-weight:700;font-size:16pt;padding-bottom:0pt;font-family:"Arial";line-height:1.15;page-break-after:avoid;orphans:2;widows:2;text-align:left}.subtitle{padding-top:0pt;color:#666666;font-size:15pt;padding-bottom:16pt;font-family:"Arial";line-height:1.15;page-break-after:avoid;orphans:2;widows:2;text-align:left}li{color:#000000;font-size:11pt;font-family:"Arial"}p{margin:0;color:#000000;font-size:11pt;font-family:"Arial"}h1{padding-top:0pt;color:#000000;font-weight:700;font-size:14pt;padding-bottom:0pt;font-family:"Arial";line-height:1.15;page-break-after:avoid;orphans:2;widows:2;text-align:left}h2{padding-top:0pt;color:#000000;font-weight:700;font-size:12pt;padding-bottom:0pt;font-family:"Arial";line-height:1.0;page-break-after:avoid;orphans:2;widows:2;text-align:left}h3{padding-top:16pt;color:#434343;font-size:14pt;padding-bottom:4pt;font-family:"Arial";line-height:1.15;page-break-after:avoid;orphans:2;widows:2;text-align:left}h4{padding-top:14pt;color:#666666;font-size:12pt;padding-bottom:4pt;font-family:"Arial";line-height:1.15;page-break-after:avoid;orphans:2;widows:2;text-align:left}h5{padding-top:12pt;color:#666666;font-size:11pt;padding-bottom:4pt;font-family:"Arial";line-height:1.15;page-break-after:avoid;orphans:2;widows:2;text-align:left}h6{padding-top:12pt;color:#666666;font-size:11pt;padding-bottom:4pt;font-family:"Arial";line-height:1.15;page-break-after:avoid;font-style:italic;orphans:2;widows:2;text-align:left}
  </style>
  <title></title>
</head>
<body class="c11">
  <p class="c6 title" id="h.xkuvp2l0zppk"><span>SCREEN –
  About</span></p>
  <hr>
  <h1 class="c6" id="h.59dk8a3j2bag"><span>The ENCODE
  Encyclopedi</span><span>a</span><sup><a href="#cmnt1" id=
  "cmnt_ref1">[a]</a></sup></h1>
  <p class="c1"><span style=
  "overflow: hidden; display: inline-block; margin: 0.00px 0.00px; border: 0.00px solid #000000; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px); width: 624.00px; height: 377.33px;">
  <img alt="" src="images/image04.png" style=
  "width: 624.00px; height: 377.33px; margin-left: 0.00px; margin-top: 0.00px; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px);"
  title=""></span><span class="c0">The ENCODE Encyclopedia
  comprises two levels of annotations. The ground level includes
  peaks and quantifications produced by uniform processing
  pipelines for individual data types. &nbsp;The integrative level
  contains annotations produced by integrating multiple data types.
  In the core of the integrative level is the Registry of candidate
  Regulatory Elements (cREs), and SCREEN is a web-based
  visualization engine specifically designed for the Registry.
  SCREEN allows users to explore cREs and investigate how these
  elements connect with other annotations in the Encyclopedia, as
  well as raw ENCODE data whenever informative.</span></p>
  <p class="c1 c4"></p>
  <p class="c1 c4"></p>
  <h1 class="c6" id="h.c764829z5hni"><span class="c9">The Registry
  of candidate Regulatory Elements</span></h1>
  <p class="c1"><span>The cREs in the Registry are the subset of
  representative DNase hypersensitivity sites (rDHSs) that are
  supported by histone modification and CTCF-binding data. We
  started with the DHSs called from 486 DNase-seq experiments in
  humans and 62 experiments in mouse using the</span>
  <span><a class="c2" href=
  "https://www.google.com/url?q=https://www.encodeproject.org/search/?type%3DExperiment%26internal_tags%3DENCYCLOPEDIAv4%26assay_slims%3DDNA%2Baccessibility%26assembly%3Dhg19%26files.file_type%3Dbed%2BbroadPeak&amp;sa=D&amp;ust=1487778763358000&amp;usg=AFQjCNExZ392tjp0AcWB2u5p5XLtiH_t-A">
  &nbsp;</a></span><span class="c3"><a class="c2" href=
  "https://www.google.com/url?q=https://www.encodeproject.org/search/?type%3DExperiment%26internal_tags%3DENCYCLOPEDIAv4%26assay_slims%3DDNA%2Baccessibility%26assembly%3Dhg19%26files.file_type%3Dbed%2BbroadPeak&amp;sa=D&amp;ust=1487778763358000&amp;usg=AFQjCNExZ392tjp0AcWB2u5p5XLtiH_t-A">Hotspot2</a></span><span>&nbsp;algorithm
  (the algorithm was created by the Stam lab and run by the ENCODE
  data coordination center) and created the list of rDHSs
  using</span> <span class="c3"><a class="c2" href=
  "https://www.google.com/url?q=https://www.ncbi.nlm.nih.gov/pubmed/22576172&amp;sa=D&amp;ust=1487778763359000&amp;usg=AFQjCNE6L2Q8deBPJ1iwURmjQeszctSYqw">
  Bedops</a></span><span>&nbsp;and a custom script by Robert
  Thurman in John Stam's lab (</span><span class="c3"><a class="c2"
  href=
  "https://www.google.com/url?q=https://github.com/Jill-Moore/ENCODE3/blob/master/Registry-of-cREs/Create-cREs.sh&amp;sa=D&amp;ust=1487778763360000&amp;usg=AFQjCNHhy-E_a6CXf_Wit8GHxik-yVVRmw">gitHub</a></span><span>).
  This script iteratively merges DHSs across all experiments and
  selects the most significant DHS (i.e., the DHS with lowest false
  discovery rate or FDR). This process continues until it finally
  results in a list of non-overlapping rDHSs representing every DHS
  (</span><span class="c13">Figure 1</span><span>). We then
  filtered these rDHSs using an FDR threshold of 0.1%. This
  pipeline resulted in 2,568,674 rDHSs in human and 1,268,875 rDHSs
  in mouse, starting from the initial ~80M DHSs in human and ~15M
  DHSs in mouse at FDR 5% (</span><span class="c13">Figure
  2</span><span>). We then asked whether each rDHS is supported by
  H3K4me3, H3K27ac, or CTCF signals. Our first step was to
  normalize signals between ChIP-seq experiments so that they were
  directly comparable. We calculated the average signal (in a 300
  bp window for DNase and CTCF and a 1000 bp window for H3K27ac and
  H3K4me3) centered on each rDHS. We then transformed these
  averages to Z-</span><span>scores</span><span class="c0">&nbsp;by
  calculating the mean and standard deviation across all
  rDHSs.</span><span style=
  "overflow: hidden; display: inline-block; margin: 0.00px 0.00px; border: 0.00px solid #000000; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px); width: 303.88px; height: 481.50px;"><img alt="Category-Flow-Chart.V4.png"
  src="images/image02.png" style=
  "width: 306.09px; height: 481.50px; margin-left: -1.11px; margin-top: 0.00px; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px);"
  title=""></span><span style=
  "overflow: hidden; display: inline-block; margin: 0.00px 0.00px; border: 0.00px solid #000000; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px); width: 325.50px; height: 206.28px;"><img alt="F2.MasterPeak.Overview.png"
  src="images/image00.png" style=
  "width: 327.50px; height: 206.28px; margin-left: -1.00px; margin-top: 0.00px; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px);"
  title=""></span></p>
  <p class="c1 c4"></p>
  <p class="c1"><span>We assigned rDHSs to activity groups using a
  classification tree (</span><span class="c13">Figure
  2</span><span class="c0">). First, we selected all rDHSs with a
  DNase max Z-score (across all cell types) of 1.64, resulting in
  1.76M (68.5%) high-signal rDHSs. Then using H3K4me3 and H3K27ac
  max Z-scores and distance from TSSs, we identified Promoter-like
  and Enhancer-like cREs. From the rDHSs that did not fall into
  either category, we identified CTCF-only cREs using a CTCF max
  Z-score &gt; 1.64. We separated the remaining rDHSs two groups
  based on whether they overlapped with peaks of active histone
  marks (H3K4me1/2/3, H3K27ac, and H3K9ac). The remaining 76,396
  rDHSs were not included in the Registry.</span></p>
  <p class="c1 c4"></p>
  <p class="c1"><span class="c0">Using this classification scheme,
  each cRE can only belong to one group (Promoter-like,
  Enhancer-like, CTCF-only, other cREs supported by active histone
  marks) and these assignments are cell type agnostic. As cREs have
  cell-type dependent activities, in the next section we describe
  cell-type-specific assignments.</span></p>
  <p class="c1 c4"></p>
  <p class="c1"><span>We analyzed the percentage of the genome
  comprised of each cRE category (</span><span class="c13">Figure
  3</span><span>). We only considered regions of the genome which
  is mappable by 36-nt long sequences in DNase-seq experiments
  (~2.65 Gb).</span> <span>In total, cREs cover 24.9%</span>
  <span class="c0">of the mappable genome.</span></p>
  <p class="c1"><span style=
  "overflow: hidden; display: inline-block; margin: 0.00px 0.00px; border: 0.00px solid #000000; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px); width: 331.72px; height: 167.50px;">
  <img alt="Mappable-Genome-Breakdown.3D.png" src=
  "images/image06.png" style=
  "width: 331.72px; height: 167.50px; margin-left: 0.00px; margin-top: 0.00px; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px);"
  title=""></span></p>
  <p class="c1 c4"></p>
  <p class="c1 c4"></p>
  <p class="c1 c4"></p>
  <h1 class="c6" id="h.tpb8vo9cxun6"><span class="c9">Annotating
  candidate Regulatory Elements</span></h1>
  <h2 class="c7 c10" id="h.fjamwvngux5n"><span class=
  "c14 c13">Genomic Context</span><span style=
  "overflow: hidden; display: inline-block; margin: 0.00px 0.00px; border: 0.00px solid #000000; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px); width: 282.12px; height: 160.50px;"><img alt="Genomic-Context-Breakdown.3D.png"
  src="images/image05.png" style=
  "width: 282.12px; height: 160.50px; margin-left: 0.00px; margin-top: 0.00px; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px);"
  title=""></span></h2>
  <p class="c7 c15"><span>Judged by</span><span>&nbsp;GENCODE V19
  annotations, 13.7</span><span>%</span><span>&nbsp;of cREs are
  proximal (±2Kb) to annotated GENCODE transcription start sites
  (TSSs) while the majority of cREs are distal from TSSs and lie in
  introns or intergenic regions. We annotated each cRE with genomic
  information including nearby genes and their expression levels,
  single nucleotide polymorphisms (SNPs), and element within the
  same topologically associated domains (TADs). These annotations
  can be found for each cRE on the "</span><span class="c5">Nearby
  Genomic Features</span><span>" tab.</span></p>
  <p class="c1 c4"></p>
  <p class="c1 c4"></p>
  <h2 class="c7 c10" id="h.ll6b7owfulql"><span>Histone
  Modifications and Transcription Factor Occupancy</span></h2>
  <p class="c7 c15"><span>In addition to chromatin accessibility
  measured by DNase-seq experiments, each</span><span class=
  "c0">&nbsp;cRE is annotated using H3K27ac and H3K4me3 levels and
  CTCF occupancy to determine their potential regulatory roles
  (e.g. enhancer-like, promoter-like, or CTCF-bound
  insulator-like). &nbsp;Our first step was to normalize signals
  between experiments so that they were directly comparable. We
  transformed the average signal across a cRE to the Z-score of
  log(signal).</span> <span style=
  "overflow: hidden; display: inline-block; margin: 0.00px 0.00px; border: 0.00px solid #000000; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px); width: 297.50px; height: 227.50px;">
  <img alt="Cell-Type-Venn.png" src="images/image03.png" style=
  "width: 297.50px; height: 227.50px; margin-left: 0.00px; margin-top: 0.00px; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px);"
  title=""></span></p>
  <p class="c7 c15 c4"></p>
  <p class="c7 c15"><span>For each cRE, we calculated the Z-score
  of log(signal) across all cell and tissue types for DNase,
  H3K4me3, H3K27ac, and CTCF. We also computed a Promoter-like and
  Enhancer-like Z-score by averaging DNase & H3K4me3 and DNase &
  H3K27ac Z-scores respectively. This scores are available under
  the "</span><span class="c5">Top Tissues</span><span>" tab.
  Additionally, we also intersect each cRE all available histone
  mark and TF ChIP-seq peaks. We display these results in the
  "</span><span class="c5">TF and His-mod
  Intersection</span><span>" Tab.</span><span class=
  "c0">&nbsp;</span><span style=
  "overflow: hidden; display: inline-block; margin: 0.00px 0.00px; border: 0.00px solid #000000; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px); width: 346.56px; height: 175.50px;"><img alt="Mouse.Celltype.Venn.png"
  src="images/image01.png" style=
  "width: 346.56px; height: 175.50px; margin-left: 0.00px; margin-top: 0.00px; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px);"
  title=""></span></p>
  <p class="c1 c4"></p>
  <h1 class="c6" id="h.tn1tekja57mh"><span class="c9">Using cREs to
  Interpret Variants Identified by Genome-wide Association
  Studies</span></h1>
  <p class="c1 c4"></p>
  <p class="c1"><span class="c13">Curating GWAS Results</span></p>
  <p class="c1"><span>Using the</span> <span class="c3"><a class=
  "c2" href=
  "https://www.google.com/url?q=https://www.ebi.ac.uk/gwas/&amp;sa=D&amp;ust=1487778763382000&amp;usg=AFQjCNG-7w0pMPUwYbfRPTIoFMGSQu2Q4g">
  NHGRI-EBI GWAS Catalog</a></span><span class="c0">, we selected
  studies with the following requirements:</span></p>
  <ol class="c19 lst-kix_i49qhjyjm8rs-0 start" start="1">
    <li class="c1 c21"><span class="c0">Study must be performed on
    single population. Mixed populations will complicated LD
    structure.</span></li>
    <li class="c1 c21"><span class="c0">Study must use subject from
    CEU (caucasian-european) population. For the time being we are
    curating other these studies as we are using population
    specific LD data. We plan to extend this other populations in
    the near future.</span></li>
  </ol>
  <p class="c1"><span class="c0">For each study, we downloaded all
  report SNPs, even those that were just under genome wide
  significance.</span></p>
  <p class="c1 c4"></p>
  <p class="c1"><span class="c8">Determining Enriched Cell
  Types</span></p>
  <p class="c1"><span>For each study, we created a control set of
  SNPs accounting for minor allele frequency (in CEU population)
  and distance from TSS using SNPs from SNP-Chip arrays. This
  method was adapted and modified from the</span> <span class=
  "c3"><a class="c2" href=
  "https://www.google.com/url?q=https://github.com/robertkleinlab/uesEnrichment&amp;sa=D&amp;ust=1487778763386000&amp;usg=AFQjCNHjXtlDa1bp07Y9Cj1o8GOBozwFKQ">
  Uncovering Enrichment through Simulation
  (UES)</a></span><span>&nbsp;method developed by the Klein Lab
  (Hayes</span> <span class="c5">et al.</span><span class=
  "c5 c13">&nbsp;</span><span>2015). For each GWAS tagged SNP we
  generated 100 matched controls and also included all SNPs in LD
  (default r</span><span class="c20">2</span><span class=
  "c0">&nbsp;&gt; 0.7).</span></p>
  <p class="c1 c4"></p>
  <p class="c1"><span>Using all GWAS SNPs + LD and Control SNPs +
  LD we then intersected the SNPs with cREs. For cell type specific
  enrichment, we required the cRE to have a Z-score of 1.64 in that
  cell type. After pruning for LD, (i.e. only counting one
  cell-type specific cRE per LD block) we used Fisher's exact test
  to determine if the GWAS SNPs were enriched in cell-type specific
  cREs. We generated results using H3K27ac Z-scores and will also
  calculate enrichment for DNase in the near future.</span></p>
  <p class="c1 c4"></p>
  <p class="c1"><span class="c8">Data in SCREEN</span></p>
  <p class="c1 c4"></p>
  <h1 class="c6" id="h.cc9waoinl4s4"><span>How to Cite the ENCODE
  Encyclopedia, the Registry of cREs, and SCREEN</span></h1>
  <ul class="c19 lst-kix_pkhhiafczwwt-0 start">
    <li class="c1 c18"><span>ENCODE Project Consortium, Bernstein
    BE, Birney E, Dunham I, Green ED, Gunter C, Snyder M. 2012. An
    integrated encyclopedia of DNA elements in the human
    genome.</span> <span class="c5">Nature</span><span class=
    "c0">&nbsp;489: 57–74.</span></li>
    <li class="c1 c18"><span>ENCODE Project Consortium, Myers RM,
    Stamatoyannopoulos J, Snyder M, Dunham I, Hardison RC,
    Bernstein BE, Gingeras TR, Kent WJ, Birney E, et al. 2011. A
    user's guide to the encyclopedia of DNA elements
    (ENCODE).</span> <span class="c5">PLoS
    Biol</span><span>&nbsp;9: e1001046.</span></li>
  </ul>
  <div>
    <p class="c15 c4 c22"></p>
  </div>
  <div class="c12">
    <p class="c7"><a href="#cmnt_ref1" id=
    "cmnt1">[a]</a><span class="c0">Add logo to bottom
    corner</span></p>
  </div>
</body>
</html>
